File: test/datgen/global_test.py
#!/usr/bin/env python

#
# (c) Copyright 2015 DESY, Eugen Wintersberger <eugen.wintersberger@desy.de>
#
# This file is part of pnitools.
#
# pnitools is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# pnitools is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with pnitools.  If not, see <http://www.gnu.org/licenses/>.
#************************************************************************
#
#  Created on: Jun 22, 2015
#      Author: Eugen Wintersberger <eugen.wintersberger@desy.de>
#

from subprocess import check_output
from subprocess import STDOUT
from subprocess import call
import unittest
import os
import pni.io.nx.h5 as nx
import numpy

class datgen_global_test(unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_return(self):
        pass

    def test_noargs_noopts(self):
        cmd = ['datgen']
        result = int(call(cmd))
        self.assertEqual(result,1)

    def test_nofunctor(self):
        cmd = ['datgen','-b100','-e200','-s2']
        self.assertEqual(int(call(cmd)),1);

    def test_wrong_functor(self):
        cmd = ['datgen','-b100','-e200','-s2','uni']
        self.assertEqual(int(call(cmd)),1);
    
    def test_missing_grid_endpoint(self):
        result = int(call(['datgen','-b100','-s2','uniform']))
        self.assertEqual(result,1)
    
    def test_missing_grid_startpoint(self):
        result = int(call(['datgen','-e100','-s2','uniform']))
        self.assertEqual(result,1)

    def test_missing_grid_steps(self):
        result = int(call(['datgen','-e100','-b2','uniform']))
        self.assertEqual(result,1)

    def test_grid_end_small_start(self):
        result = int(call(['datgen','-b100','-e2','-s1','uniform']))
        self.assertEqual(result,1)

    def test_grid(self):
        cmd = ['datgen','-b100','-e200','-s10','--show-grid','uniform']
        result = check_output(cmd)
        data = numpy.fromstring(result,sep=" ").reshape(10,2)
        xdata = data[:,0]
        delta = (200.-100.)/(10.-1.)
        x = 100
        for xcli in xdata:
            self.assertAlmostEqual(xcli,x,8)
            x += delta




if __name__ == "__main__":
    unittest.main()

File: test/datgen/global_test.cpp
//
// (c) Copyright 2017 DESY
//
// This file is part of pniio.
//
// pniio is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// pniio is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with pniio.  If not, see <http://www.gnu.org/licenses/>.
// 
// ===========================================================================
//
// Author: Eugen Wintersberger <eugen.wintersberger@desy.de>
// Created: Fri 23 Jun, 2017
//
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE Global tests for datgen
#include <boost/test/unit_test.hpp>
#include <boost/filesystem.hpp>
#include <iostream>
#include <vector>

#include <test/config.hpp>
#include <test/command_runner.hpp>
#include "datgen_fixture.hpp"

using namespace pni::core;

BOOST_FIXTURE_TEST_SUITE(datgen_global_test,datgen_fixture)

    BOOST_AUTO_TEST_CASE(test_noargs_nopts)
    {
        BOOST_CHECK_EQUAL(runner({}),1);
    }

    BOOST_AUTO_TEST_CASE(test_nofunctor)
    {
        BOOST_CHECK_EQUAL(runner({"-b100","-e200","-s2"}),1);
    }

    BOOST_AUTO_TEST_CASE(test_wrong_functor)
    {
        BOOST_CHECK_EQUAL(runner({"-b100","-e200","-s2","uni"}),1);
    }

    BOOST_AUTO_TEST_CASE(test_missing_grid_endpoint)
    {
        BOOST_CHECK_EQUAL(runner({"-b100","-s2","uniform"}),1);
    }

    BOOST_AUTO_TEST_CASE(test_missing_grid_startpoint)
    {
        BOOST_CHECK_EQUAL(runner({"-e100","-s2","uniform"}),1);
    }

    BOOST_AUTO_TEST_CASE(test_missing_grid_steps)
    {
        BOOST_CHECK_EQUAL(runner({"-e100","-b2","uniform"}),1);
    }

    BOOST_AUTO_TEST_CASE(test_grid_end_small_start)
    {
        BOOST_CHECK_EQUAL(runner({"-b100","-e2","-s1","uniform"}),1);
    }

    BOOST_AUTO_TEST_CASE(test_grid)
    {
        BOOST_CHECK_EQUAL(runner({"-b100","-e200","-s10","--show-grid","uniform"}),0);
        read_results();
        BOOST_CHECK_EQUAL(x.size(),10);
        BOOST_CHECK_EQUAL(y.size(),10);
        float64 x_ref=100.; 
        for(auto iter_x = x.begin();iter_x != x.end(); ++iter_x,x_ref+=(200.-100.)/(10.-1.))
        {
            BOOST_CHECK_CLOSE(*iter_x,x_ref,0.001);
        }
        
        
    }



BOOST_AUTO_TEST_SUITE_END()
